### ISSUE TITLE
Tensorboard summary of learning rate

### ISSUE THREAD

**Issue by tingkai-zhang** [allenai/allennlp#2388]
Hi, i am using cosine with restart with the following configuration:
```
    "learning_rate_scheduler": {
       "type":"cosine",
       "t_initial": 500,
       "t_mul": 1.5,
       "eta_min": 0.0,
       "eta_mul": 0.7

    },
    "optimizer": {
      "type": "adam",
      "lr": 3e-5
    }
```
The training has 1.2k steps per epoch.
I think what that the tensorboard summary shows that there is minor change of learning rate(0.001e-5) and not giving a cosine-like shape curve? Does the graph make sense?
![image](https://user-images.githubusercontent.com/20318562/51363397-c5162680-1aa5-11e9-9c96-bfddeec51170.png)

**Comment by epwalsh** [comment]
This learning rate scheduler only updates at the end of each epoch, which is why it is flat through the 1.2K steps within each epoch. Does that answer your question? (there's no particular reason why it doesn't update "smoothly" instead of only at the end of each epoch, other than it was just easier to implement, TBH)

**Comment by brendan-ai2** [comment]
Thanks, @epwalsh!

@tingkai-zhang , If desired you can update per-step by making your own learning rate scheduler. See, for instance, https://github.com/allenai/allennlp/blob/master/allennlp/training/learning_rate_schedulers.py#L138.

**Comment by tingkai-zhang** [comment]
@epwalsh  Hi, thanks for answering! 

> This learning rate scheduler only updates at the end of each epoch, which is why it is flat through the 1.2K steps within each epoch.

If i understand it properly, you mean the `lr` is updated each iteration but only logged to tensorboard each epoch, am i right?

If the updated learning rate is only logged at the end of each epoch, the learning rate should be `eta_min` which should not be `3e-5` in my case, right? 


@brendan-ai2  Thanks for your advice! I will look into that! 
When i use slanted triangular scheduler, I also observe the following learning rate on tensorboard is (probably) not we want?

There is a cut of line on the right side because I got such screenshot while the model is still in training

![image](https://user-images.githubusercontent.com/20318562/51420547-fbfd4280-1b60-11e9-9f4b-2e92e9517253.png)

![image](https://user-images.githubusercontent.com/20318562/51420537-edaf2680-1b60-11e9-9275-886bc9d5539f.png)


I thought the unfreezed layer would have the learning rate graph as following till the end of training (not till the end of an epoch).
![image](https://user-images.githubusercontent.com/20318562/51420463-31edf700-1b60-11e9-9b17-8c08be43a1ec.png)

But the graph of a unfreezed layer i got shows that, it first does a 'small triangular scheduler` within one epoch and does a 'big triangular scheduler' after unfreezing all layers


The config i use is 

```
    "learning_rate_scheduler": {
      "type": "slanted_triangular",
      "num_epochs": 10,
      "num_steps_per_epoch": 1522,
      "gradual_unfreezing": true,
      "discriminative_fine_tuning": true,
      "decay_factor": 0.7
    },
    "optimizer": {
      "type": "adam",
      "lr": 5e-5,
      "parameter_groups": [
        [['layer1'], {"lr": 0}],
        [['layer2'], {}],
        [['layer3'], {}],
        [['layer4'], {}],
        [['clf_head'], {}],
      ]
    },
```

**Comment by sethah** [comment]
From the slanted triangular learning rate schedule doc:

```
If we gradually unfreeze, then in the first epoch of training, only the top
layer is trained; in the second epoch, the top two layers are trained, etc.
During freezing, the learning rate is increased and annealed over one epoch.
After freezing finished, the learning rate is increased and annealed over
the remaining training iterations.
```

So, it's correct. I think it's a bit fuzzy how to implement the one-cycle policy when you are unfreezing some of the layers. If you look at some of the FastAI lessons, this is essentially how they do it: unfreeze some params, run one cycle, unfreeze more, run one cycle, etc... So I believe the implementation was meant to match that. 

Regarding the cosine scheduler, I believe it is only updated once per epoch. The scheduler has no `step_batch` function, and according to the docs of `trainer`:

```
learning_rate_scheduler : ``LearningRateScheduler``, optional (default = None)
            If specified, the learning rate will be decayed with respect to
            this schedule at the end of each epoch (or batch, if the scheduler implements
            the ``step_batch`` method)
```

You can control how often the learning rate gets logged to tensorboard, btw using the `should_log_learning_rate` and `summary_interval` parameters of the `trainer` class.

**Comment by tingkai-zhang** [comment]
@sethah  Thank you for replying! 
For the slanted triangular one, I have figured it out with your explanation. Thank you!

For the cosine with restart one, here is why I asked such problem.
The doc says `t_initial` is `The number of iterations within the first cycle.`
But it is actually means the `The number of **epochs** within the first cycle.`
That why the graph shows that the lr is unchanged during the first 500 **epochs** when I set `t_initial` to 500 because doc says it is the number of **iterations** within the first cycle.

My first question has nothing to do with updating per epoch/step. In the previous conversation, I pointed out even if you take it as update per epoch, the graph is not correct if you think `t_initial` is the number of iterations. But now I know the implementation is correct. 

Maybe I should pull a request to fix the doc?
